components/tidb_query/src/rpn_expr/impl_string.rs

```diff
@@ -931,6 +950,52 @@ mod tests {
         }
     }
 
+    #[test]
+    fn test_upper() {
+        let cases = vec![
+            (Some(b"hello".to_vec()), Some(b"HELLO".to_vec())),
+            (Some(b"123".to_vec()), Some(b"123".to_vec())),
+            (
+                Some("café".as_bytes().to_vec()),
+                Some("CAFÉ".as_bytes().to_vec()),
+            ),
+            (
+                Some("数据库".as_bytes().to_vec()),
+                Some("数据库".as_bytes().to_vec()),
+            ),
+            (
+                Some("ночь на окраине москвы".as_bytes().to_vec()),
+                Some("НОЧЬ НА ОКРАИНЕ МОСКВЫ".as_bytes().to_vec()),
+            ),
+            (
+                Some("قاعدة البيانات".as_bytes().to_vec()),
+                Some("قاعدة البيانات".as_bytes().to_vec()),
+            ),
+            (None, None),
+        ];
+
+        for (arg, exp) in cases {
+            // Test binary string case
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg.clone())
+                .evaluate(ScalarFuncSig::Upper)
+                .unwrap();
+            assert_eq!(output, exp);
+
+            // Test non-binary string case
+            let mut ft = FieldType::default();
+            ft.as_mut_accessor()
+                .set_tp(FieldTypeTp::String)
+                .set_collation(Collation::Binary);
+            let output = RpnFnScalarEvaluator::new()
+                .push_param(arg.clone())
+                .evaluate_raw(ft, ScalarFuncSig::Upper)
```

Comment:
I think you can use `.return_field_type` to set the return field type, instead of using `evaluate_raw`.

Reply:
Solved:)